File: MyDailyBeat/MyDailyBeat/EVCHealthPortalTableViewController.m
//
//  EVCHealthPortalTableViewController.m
//  MyDailyBeat
//
//  Created by Virinchi Balabhadrapatruni on 3/26/16.
//  Copyright © 2016 eVerveCorp. All rights reserved.
//

#import "EVCHealthPortalTableViewController.h"
#import "Constants.h"
#import "HealthDatabase.h"
#import <AHKActionSheet.h>

@interface EVCHealthPortalTableViewController ()

@end

@implementation EVCHealthPortalTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.healthPortals = [[NSMutableArray alloc] initWithArray:[[HealthDatabase database] healthPortals]];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.healthPortals count] + 1;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"CELL"];
    if (indexPath.row == [self.healthPortals count]) {
        cell.textLabel.text = @"Add Portal";
        cell.imageView.image = [EVCCommonMethods imageWithImage:[UIImage imageNamed:@"plus-512.png"] scaledToSize:CGSizeMake(30, 30)];
    } else {
        cell.textLabel.text = ((HealthInfo * )[self.healthPortals objectAtIndex:indexPath.row]).URL;
        cell.imageView.image = nil;
    }
    
    
    
    
    // Configure the cell...
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
    if (indexPath.row == [self.healthPortals count]) {
        DLAVAlertView *alert = [[DLAVAlertView alloc] initWithTitle:@"Enter new health portal." message:@"" delegate:nil cancelButtonTitle:@"Cancel" otherButtonTitles:@"OK", nil];
        [alert addTextFieldWithText:@"" placeholder:@""];
        [alert showWithCompletion:^(DLAVAlertView *alertView, NSInteger buttonIndex) {
            switch (buttonIndex) {
                case 0:
                    // cancel
                    break;
                case 1: {
                    // ok
                    NSString *text = [alertView textFieldTextAtIndex:0];
                    HealthInfo *portal = [[HealthInfo alloc] initWithUniqueId:0 URL:text logoURL:@""];
                    [[HealthDatabase database] insertIntoDatabase:portal orPrescriptionProvider:nil];
                    self.healthPortals = [[NSMutableArray alloc] initWithArray:[[HealthDatabase database] healthPortals]];
                    [self.tableView reloadData];
                }
                    
                default:
                    break;
            }
        }];
    } else {
        [self popupActionMenu:indexPath.row];
    }
    
}

- (void) openURLinBrowser: (NSString *) url {
    NSString *fullURL = [NSString stringWithFormat:@"%@", url];
    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:fullURL]];
    
}

- (void) popupActionMenu: (int) row {
    AHKActionSheet *sheet = [[AHKActionSheet alloc] initWithTitle:@""];
    [sheet addButtonWithTitle:@"Open App" type:AHKActionSheetButtonTypeDefault handler:^(AHKActionSheet *actionSheet) {
        HealthInfo *obj = [self.healthPortals objectAtIndex:row];
        [self openURLinBrowser:obj.URL];
    }];
    [sheet addButtonWithTitle:@"Set as My Health Portal" type:AHKActionSheetButtonTypeDefault handler:^(AHKActionSheet *actionSheet) {
        HealthInfo *obj = [self.healthPortals objectAtIndex:row];
        NSData *encoded = [NSKeyedArchiver archivedDataWithRootObject:obj];
        [[NSUserDefaults standardUserDefaults] setObject:encoded forKey:@"myHealthPortal"];
    }];
    [sheet show];
}

@end

File: MDB_Swift/MyDailyBeat/MyDailyBeat/src/EVCHealthPortalTableViewController.swift

//
//  EVCHealthPortalTableViewController.swift
//  MyDailyBeat
//
//  Created by Virinchi Balabhadrapatruni on 3/26/16.
//  Copyright © 2016 eVerveCorp. All rights reserved.
//
import UIKit
import API
class EVCHealthPortalTableViewController: UITableViewController {
    var healthPortals = [HealthInfo]()


    override func viewDidLoad() {
        super.viewDidLoad()
        self.healthPortals = DataManager.getHealthPortals()
    }

    
// MARK: - Table view data source

override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.healthPortals.count + 1
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = UITableViewCell(style: .default, reuseIdentifier: "CELL")
        if indexPath.row == self.healthPortals.count {
            cell.textLabel?.text = "Add Portal"
            cell.imageView?.image = EVCCommonMethods.image(with: UIImage(named: "plus-512.png")!, scaledTo: CGSize(width: CGFloat(30), height: CGFloat(30)))
        }
        else {
            cell.textLabel?.text = self.healthPortals[indexPath.row].url
            cell.imageView?.image = nil
        }
        // Configure the cell...
        return cell
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        self.tableView.deselectRow(at: indexPath, animated: true)
        if indexPath.row == self.healthPortals.count {
            let alert = UIAlertController(title: "Enter new health portal.", message: "Enter the link to the health portal you wish to add.", preferredStyle: .alert)
            alert.addTextField(configurationHandler: { (textField) in
                textField.text = ""
                textField.placeholder = ""
            })
            let cancel = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
            let ok = UIAlertAction(title: "OK", style: .default, handler: { (action) in
                
                if let text: String = alert.textFields![0].text, !self.healthPortals.contains(where: { (info) -> Bool in
                    return info.url == text
                }){
                    let fullURL: String
                    if text.hasPrefix("http://") || text.hasPrefix("https://") {
                        fullURL = "\(text)"
                    } else {
                        fullURL = "http://\(text)"
                    }
                    if let url = URL(string: fullURL), isURLValid(url) {
                        let portal = HealthInfo(uniqueId: 0, url: text, logoURL: "")
                        DataManager.insertHealthPortal(portal)
                        self.healthPortals = DataManager.getHealthPortals()
                    }
                }
                self.tableView.reloadData()
            })
            alert.addAction(cancel)
            alert.addAction(ok)
            self.present(alert, animated: true, completion: nil)
        }
        else {
            self.popupActionMenu(indexPath.row)
        }
    }

    func openURLinBrowser(_ url: String) {
        let fullURL: String
        if url.hasPrefix("http://") || url.hasPrefix("https://") {
            fullURL = "\(url)"
        } else {
            fullURL = "http://\(url)"
        }
        UIApplication.shared.open(URL(string: fullURL)!, options: [:], completionHandler: nil)
    }

    func popupActionMenu(_ row: Int) {
        let sheet = UIAlertController(title: "", message: "", preferredStyle: .actionSheet)
        let browserAction = UIAlertAction(title: "Open Site", style: .default) { (action) in
            let obj: HealthInfo? = self.healthPortals[row]
            self.openURLinBrowser((obj?.url)!)
        }
        let addAction = UIAlertAction(title: "Set as My Health Portal", style: .default) { (action) in
            let obj: HealthInfo = self.healthPortals[row]
            let encoded = NSKeyedArchiver.archivedData(withRootObject: obj)
            UserDefaults.standard.set(encoded, forKey: "myHealthPortal")
        }
        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
        sheet.addAction(browserAction)
        sheet.addAction(addAction)
        sheet.addAction(cancelAction)
        self.present(sheet, animated: true, completion: nil)
    }
}
